compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -940,47 +848,31 @@ mod test {
 
     #[test]
     fn remove_inc_rcs_that_are_never_mutably_borrowed() {
-        // acir(inline) fn main f0 {
-        //     b0(v0: [Field; 2]):
-        //       inc_rc v0
-        //       inc_rc v0
-        //       inc_rc v0
-        //       v2 = array_get v0, index u32 0
-        //       inc_rc v0
-        //       return v2
-        //   }
-        let main_id = Id::test_new(0);
-
-        // Compiling main
-        let mut builder = FunctionBuilder::new("main".into(), main_id);
-        let v0 = builder.add_parameter(Type::Array(Arc::new(vec![Type::field()]), 2));
-        builder.increment_array_reference_count(v0);
-        builder.increment_array_reference_count(v0);
-        builder.increment_array_reference_count(v0);
-
-        let zero = builder.numeric_constant(0u128, Type::unsigned(32));
-        let v2 = builder.insert_array_get(v0, zero, Type::field());
-        builder.increment_array_reference_count(v0);
-        builder.terminate_with_return(vec![v2]);
-
-        let ssa = builder.finish();
+        let src = "
+            acir(inline) fn main f0 {
+              b0(v0: [Field; 2]):
+                inc_rc v0
+                inc_rc v0
+                inc_rc v0
+                v2 = array_get v0, index u32 0 -> Field
```

Comment:
Could the parser or the builder plausibly figure out that an `array_get` on `v0: [Field; 2]` means that `v2` is of type `Field` without having to add `-> Field`?

Reply:
It's a great question, and it's something I wondered at least twice when writing this PR.

Consider this code:

```noir
fn main(x: Field) -> pub Field {
    let array = [(1, true)];
    let value = array[x];
    value.0
}
```

When accessing an array index, when the array is composite, what Noir will end up doing is accessing each element in the composite type in turn by shifting an index. Here's the SSA the above initially produces:

```
acir(inline) fn main f0 {
  b0(v0: Field):
    inc_rc [Field 1, u1 1] of (Field, u1)
    v4 = cast v0 as u32
    v6 = mul v4, u32 2
    v7 = array_get [Field 1, u1 1] of (Field, u1), index v6 -> Field
    v9 = add v6, u32 1
    v10 = array_get [Field 1, u1 1] of (Field, u1), index v9 -> u1
    return v7
}
```

that's with the type annotations that are added in this PR. Two things to note:
1. An array has all the values flattened, so `1` is on position 0, `true` is on position 1.
2. `array_get` works on these indices, not on the actual indices on the Noir side, so accessing 0 will get you 1, accessing 1 will get you `true`
3. When we do `array_get` we don't know which type will get. Well, actually we can guess because the first one will be one the first type, the second one on the second type, etc... but I didn't want to make the parsing too complex so I decided to just put the expected types in the code. Maybe it also makes it easier to understand the code.

Reply:
Amazing, thanks for that explanation! I agree in that case it would be difficult to figure out what `index v9` might be, and also agree that it makes it easier to follow 👍